# Ejercicios ensembling
En este ejercicio vas a realizar prediciones sobre un dataset de ciudadanos indios diabéticos. Se trata de un problema de clasificación en el que intentaremos predecir 1 (diabético) 0 (no diabético).

### 1. Carga las librerias que consideres comunes al notebook

In [22]:
import pandas as pd
from sklearn.ensemble import BaggingClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import KFold, cross_val_score
from xgboost import XGBClassifier
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.ensemble import BaggingClassifier, RandomForestClassifier, AdaBoostClassifier, GradientBoostingClassifier

### 2. Lee los datos de [esta direccion](https://raw.githubusercontent.com/jbrownlee/Datasets/master/pima-indians-diabetes.data.csv)
Los nombres de columnas son:
```Python
names = ['preg', 'plas', 'pres', 'skin', 'test', 'mass', 'pedi', 'age', 'class']
```

In [23]:
df = pd.read_csv(r"https://raw.githubusercontent.com/jbrownlee/Datasets/master/pima-indians-diabetes.data.csv")

In [24]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 767 entries, 0 to 766
Data columns (total 9 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   6       767 non-null    int64  
 1   148     767 non-null    int64  
 2   72      767 non-null    int64  
 3   35      767 non-null    int64  
 4   0       767 non-null    int64  
 5   33.6    767 non-null    float64
 6   0.627   767 non-null    float64
 7   50      767 non-null    int64  
 8   1       767 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


In [25]:
df.head()

,6,148,72,35,0,33.6,0.627,50,1
0,1,85,66,29,0,26.6,0.351,31,0
1,8,183,64,0,0,23.3,0.672,32,1
2,1,89,66,23,94,28.1,0.167,21,0
3,0,137,40,35,168,43.1,2.288,33,1
4,5,116,74,0,0,25.6,0.201,30,0


In [26]:
df.columns = ['preg', 'plas', 'pres', 'skin', 'test', 'mass', 'pedi', 'age', 'class'] #Cambiamos los nombres de las columnas

In [27]:
df.head()

,preg,plas,pres,skin,test,mass,pedi,age,class
0,1,85,66,29,0,26.6,0.351,31,0
1,8,183,64,0,0,23.3,0.672,32,1
2,1,89,66,23,94,28.1,0.167,21,0
3,0,137,40,35,168,43.1,2.288,33,1
4,5,116,74,0,0,25.6,0.201,30,0


### 3. Bagging
Para este apartado tendrás que crear un ensemble utilizando la técnica de bagging ([BaggingClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.BaggingClassifier.html)), mediante la cual combinarás 100 [DecisionTreeClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html). Recuerda utilizar también [cross validation](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.KFold.html) con 10 kfolds.

**Para este apartado y siguientes, no hace falta que dividas en train/test**, por hacerlo más sencillo. Simplemente divide tus datos en features y target.

Establece una semilla

In [28]:
# 1. Separar features y target
X = df.drop('class', axis=1)
y = df['class']

# 2. Configurar la semilla y el KFold
kfold = KFold(n_splits=10, random_state=29, shuffle=True)

# 3. Definir el modelo base (Decision Tree)
cart = DecisionTreeClassifier()

# 4. Crear el ensemble con Bagging (100 árboles)
num_trees = 100
model = BaggingClassifier(estimator=cart, n_estimators=num_trees, random_state=29)

# 5. Evaluar con Cross Validation
results = cross_val_score(model, X, y, cv=kfold)

print(f"Precisión media: {results.mean():.3f} (Desviación: {results.std():.3f})")

Precisión media: 0.761 (Desviación: 0.031)


### 4. Random Forest
En este caso entrena un [RandomForestClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html) con 100 árboles y un `max_features` de 3. También con validación cruzada

In [29]:
# Definir el modelo RandomForest
model_rf = RandomForestClassifier(n_estimators=100, max_features=3, random_state=29)

# Evaluar con Cross Validation (usando el kfold definido anteriormente)
results_rf = cross_val_score(model_rf, X, y, cv=kfold)

print(f"Precisión media (RF): {results_rf.mean():.3f} (Desviación: {results_rf.std():.3f})")

Precisión media (RF): 0.757 (Desviación: 0.030)


### 5. AdaBoost
Implementa un [AdaBoostClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.AdaBoostClassifier.html) con 30 árboles.

In [30]:
#Definir el modelo AdaBoost
model_ada = AdaBoostClassifier(n_estimators=30, random_state=29)

# Evaluar con Cross Validation (usando el kfold de 10 particiones)
results_ada = cross_val_score(model_ada, X, y, cv=kfold)

print(f"Precisión media (AdaBoost): {results_ada.mean():.3f} (Desviación: {results_ada.std():.3f})")

Precisión media (AdaBoost): 0.755 (Desviación: 0.041)


### 6. GradientBoosting
Implementa un [GradientBoostingClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.GradientBoostingClassifier.html) con 100 estimadores

In [31]:
# Definir el modelo Gradient Boosting
model_gb = GradientBoostingClassifier(n_estimators=100, random_state=29)

# Evaluar con Cross Validation
results_gb = cross_val_score(model_gb, X, y, cv=kfold)

print(f"Precisión media (Gradient Boosting): {results_gb.mean():.3f} (Desviación: {results_gb.std():.3f})")

Precisión media (Gradient Boosting): 0.765 (Desviación: 0.043)


### 7. XGBoost
Para este apartado utiliza un [XGBoostClassifier](https://docs.getml.com/latest/api/getml.predictors.XGBoostClassifier.html) con 100 estimadores. XGBoost no forma parte de la suite de modelos de sklearn, por lo que tendrás que instalarlo con pip install

In [32]:
# Definir el modelo XGBoost
model_xgb = XGBClassifier(n_estimators=100, random_state=29, eval_metric='logloss')

# Evaluar con Cross Validation
results_xgb = cross_val_score(model_xgb, X, y, cv=kfold)

print(f"Precisión media (XGBoost): {results_xgb.mean():.3f} (Desviación: {results_xgb.std():.3f})")

Precisión media (XGBoost): 0.727 (Desviación: 0.045)


### 8. Primeros resultados
Crea un dataframe con los resultados y sus algoritmos, ordenándolos de mayor a menor

In [33]:
# Creamos el diccionario con los datos recopilados
resultados_data = {
    'Algoritmo': ['Bagging', 'Random Forest', 'AdaBoost', 'Gradient Boosting', 'XGBoost'],
    'Precisión Media': [0.761, 0.757, 0.755, 0.765, 0.727],
    'Desviación': [0.031, 0.030, 0.041, 0.043, 0.045]
}

# Creamos el DataFrame
df_resultados = pd.DataFrame(resultados_data)

# Ordenamos de mayor a menor precisión
df_resultados = df_resultados.sort_values(by='Precisión Media', ascending=False).reset_index(drop=True)

In [34]:
df_resultados.head()

,Algoritmo,Precisión Media,Desviación
0,Gradient Boosting,0.765,0.043
1,Bagging,0.761,0.031
2,Random Forest,0.757,0.030
3,AdaBoost,0.755,0.041
4,XGBoost,0.727,0.045


### 9. Hiperparametrización
Vuelve a entrenar los modelos de nuevo, pero esta vez dividiendo el conjunto de datos en train/test y utilizando un gridsearch para encontrar los mejores hiperparámetros.

In [35]:
# 1. Preparación de datos
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=7)

# 2. Definición de modelos y sus mallas de parámetros (Grids)
models_and_grids = [
    ('Bagging', BaggingClassifier(random_state=7), {
        'n_estimators': [50, 100, 200]
    }),
    ('Random Forest', RandomForestClassifier(random_state=7), {
        'n_estimators': [100, 200],
        'max_features': [2, 3, 5],
        'max_depth': [None, 5, 10]
    }),
    ('AdaBoost', AdaBoostClassifier(random_state=7), {
        'n_estimators': [30, 50, 100],
        'learning_rate': [0.01, 0.1, 1.0]
    }),
    ('Gradient Boosting', GradientBoostingClassifier(random_state=7), {
        'n_estimators': [100, 200],
        'learning_rate': [0.01, 0.05, 0.1],
        'max_depth': [3, 5]
    }),
    ('XGBoost', XGBClassifier(random_state=7, eval_metric='logloss'), {
        'n_estimators': [50, 100],
        'learning_rate': [0.01, 0.05, 0.1],
        'max_depth': [3, 5]
    })
]

# 3. Ejecución del Grid Search
mejores_modelos = []

for nombre, modelo, params in models_and_grids:
    grid = GridSearchCV(modelo, params, cv=10, scoring='accuracy', n_jobs=-1)
    grid.fit(X_train, y_train)
    
    mejores_modelos.append({
        'Algoritmo': nombre,
        'Mejor Score (CV)': grid.best_score_,
        'Mejor Params': grid.best_params_,
        'Test Score': grid.score(X_test, y_test) # Evaluación final con datos nunca vistos
    })

# 4. Resultados finales
df_tuning = pd.DataFrame(mejores_modelos).sort_values(by='Mejor Score (CV)', ascending=False)
print(df_tuning[['Algoritmo', 'Mejor Score (CV)', 'Test Score']])

           Algoritmo  Mejor Score (CV)  Test Score
1      Random Forest          0.763432    0.792208
3  Gradient Boosting          0.761766    0.785714
4            XGBoost          0.750397    0.792208
2           AdaBoost          0.750370    0.785714
0            Bagging          0.742253    0.785714


### 10. Conclusiones finales

*   **Modelo Óptimo:** El **Random Forest** se consolida como el mejor modelo tras la hiperparametrización, alcanzando una precisión del **79.2%** en el conjunto de test.
*   **Relevancia del Tuning:** La optimización mediante *Grid Search* fue fundamental, especialmente para **XGBoost**, que logró corregir su tendencia al sobreajuste inicial e igualar el mejor desempeño en test.
*   **Capacidad de Generalización:** El hecho de que el *Test Score* sea superior al *CV Score* en todos los modelos confirma que los ensembles han aprendido patrones generales y son robustos ante datos no vistos.
*   **Estabilidad de los Ensembles:** Todas las técnicas basadas en árboles (Bagging, Boosting y RF) demostraron ser consistentes, manteniéndose en un rango de éxito del 78-79%, lo que sugiere que se ha alcanzado un techo de rendimiento sólido para este dataset.